moon/controller_excavator_round2.py

```diff
@@ -1,18 +1,207 @@
+                    in_exception = False
+                    while len(queue) > 0:
+                        one_step = queue.pop(0)
+                        move, value = one_step
+                        if move == "in_exception":
+                            in_exception = value
+                            continue
+                        starting_pose = self.last_position
+                        starting_yaw = self.yaw
+                        if not in_exception:
+                            self.virtual_bumper = VirtualBumper(timedelta(seconds=20), 0.1)
+                            try:
+                                with LidarCollisionMonitor(self):
+                                    make_unprotected_move(one_step)
+                            except (VirtualBumperException, LidarCollisionException) as e:
+                                in_exception = True
+                                self.virtual_bumper = None
+                                print(self.time, "excavator_controller exception in %s: %s" % (move, str(e)))
+                                traceback.print_exc()
+                                recovery_queue = []
+                                if move == "go_straight":
+                                    distance_covered = pose_distance(starting_pose, self.last_position)
+                                    print(self.time, "excavator_controller exception: distance covered so far: %f" % distance_covered)
+                                    recovery_queue.append(("turn", math.radians(90)))
+                                    recovery_queue.append(("go_straight", -1))
+                                    recovery_queue.append(("go_straight", 5.0))
+                                    recovery_queue.append(("in_exception", False))
+                                    recovery_queue.append(("turn", math.radians(-90)))
+                                    recovery_queue.append(("go_straight", 9.0))
+                                    recovery_queue.append(("turn", math.radians(-90)))
+                                    recovery_queue.append(("go_straight", 5.0))
+                                    recovery_queue.append(("turn", math.radians(90)))
+                                    recovery_queue.append(("go_straight", value - distance_covered - 8))
+                                else:
+                                    recovery_queue.append(("go_straight", -1))
+                                    recovery_queue.append(("turn", math.radians(value - (self.yaw - starting_yaw))))
+                                    recovery_queue.append(("in_exception", False))
+                                    recovery_queue.append(("go_straight", 1))
+                                queue = recovery_queue + queue
+
+                        else:
+                            make_unprotected_move(one_step)
+
+                distance, index = spatial.KDTree(vol_list).query([rx,ry])
```

Comment:
:) it is bit crazy to build tree to once query the nearest(?) obstacle, but it is "one-liner" ... why not :)

Reply:
You just watch once I replace it with solving the traveling salesman problem